Redesign PrivCount Handshake to use HMAC

teor2345 opened the pull request:
PrivCount uses a HMAC-SHA256-based handshake to verify that client and server both know a shared secret key, without revealing the key on the wire.

While a HMAC is a proven cryptographic primitive, this particular construction has not been reviewed by a cryptographer, nor has this particular implementation undergone cryptographic review. It likely contains at least one bug, and any bugs might affect security.

See #117 for my notes about the protocol handshake design, and an example handshake run.
See 	62a09db for documentation on the keys created by PrivCount, including the secret handshake key.

robgjansen commented on privcount/data_collector.py:
```diff
@@ -258,6 +262,15 @@ def refresh_config(self):
+        '''
+        # The secret handshake path should be loaded (or assigned a default)
+        # whenever the config is loaded
+        return self.config.get('secret_handshake')
```
Is it possible that `self.config` might be `None` at this point? If so, this line would cause a crash.

robgjansen commented on privcount/share_keeper.py:
```diff
@@ -231,3 +235,12 @@ def refresh_config(self):
+        '''
+        # The secret handshake path should be loaded (or assigned a default)
+        # whenever the config is loaded
+        return self.config.get('secret_handshake')
```
Is it possible that `self.config` might be `None` at this point? If so, this line would cause a crash.

robgjansen commented on privcount/tally_server.py:
```diff
@@ -324,6 +334,15 @@ def refresh_config(self):
+        '''
+        # The secret handshake path should be loaded (or assigned a default)
+        # whenever the config is loaded
+        return self.config.get('secret_handshake')
```
Is it possible that `self.config` might be `None` at this point? If so, this line would cause a crash.

teor2345 replied:
See above.

teor2345 commented:
I have refactored this common code into classes in util, which performs part of the work I intended to do in #121. Perhaps it should get its own file eventually, see #130.

I also refactored some other common code, making progress towards #121.

It also looks like I misspelt hmac in the branch name. Oops. (I don't intend to fix the branch name.)

robgjansen commented:
Sorry, didn't mean for you to do work for another ticket in this one! Looks like a great start, though. Yeah, I think a `node.py` class makes sense for common node functionality.
